Consistent representation of datetimes and use of calendar attribute (#641)
cf-convention/cf-conventions: ch04.adoc

UDUNITS defines a **`minute`** as 60 **`seconds`**, an **`hour`** as 3600 **`seconds`** and a **`day`** as 86400 **`seconds`**, consistent with the <<SI>> definitions of these non-SI units.
These are fixed units of measure.
When a leap second is inserted into UTC, the minute, hour and day affected differ by one second from their usual durations according to clock time, but the units of **`minute`**, **`hour`** and **`day`** do not.

> To avoid mistakes and confusion, it is therefore recommended that the second is used as a unit in the **`utc`** calendar (<<calendar>>).

Reviewer: ```suggestion
To avoid mistakes and confusion, the second is therefore the only recommended unit in the **`utc`** calendar (<<calendar>>).
```

* ±__nn__, the hour alone, e.g. **`-06`**, **`+02`**, **`+11`**. This format is sufficient for many time zones. Zero hours must have a positive sign e.g. **`+00`**, which means the same as **`Z`**.

* ±__nn__:__nn__, the hour and minute, separated by a colon **`:`**, e.g. **`-10:00`**, **`-03:30`**, **`+05:30`**, **`+12:45`**.

The date and time parts are separated by a single space character, which may be replaced by the letter **`T`** to align more closely to ISO 8601 and RFC 3339.
(Note, however, that legacy software applications may not be able to process the letter **`T`**.)
The space between the time and __Z__ may be omitted.
Spaces are otherwise not allowed in the format.

The default datetime string format can represent datetimes in all calendars defined in these conventions.
In this context, the default datetime format in these conventions extends the allowable set of datetime strings as defined by ISO 8601 and RFC 3339, given that these two standards only recognize the Gregorian calendar.

> Representing datetime strings in CF calendars that do not follow the Gregorian calendar (in particular the `julian` calendar and all model calendars) may fail with software libraries implementing ISO 8601 and/or RFC 3339 standards; in these cases the use of tailor-made software libraries for the CF calendars may be required.

Reviewer: ```suggestion
Representing datetime strings in CF calendars that do not follow the Gregorian calendar (in particular the **`julian`** calendar and all the model-only calendars) may fail with software libraries implementing ISO 8601 and/or RFC 3339 standards; in these cases the use of tailor-made software libraries for the CF calendars may be required.
```

| **`standard`** since CF 1.13 | no | some
| **`julian`** | 365/366 | Julian | no | all | 0001-01-01 | +∞
| **`proleptic_gregorian`** | 365/366 | Gregorian | no | all | -∞ | +∞

> | **`no_leap`**/**`365_day`** | 365 | never | no | all | -∞ | +∞

Reviewer: Thanks for spotting and correcting `no_leap`

variables:
  double time(time) ;
    time:long_name = "time" ;

>     time:units = "days since 0001-07-15" ;

Reviewer: I think we should leave the year as `1` because we can't know how many leading zeroes would be appropriate in this idealised calendar.

Author: This would then be inconsistent with the text in the (new) sub-section of "Datetime string format" in section 4.4.2, line 253 in this file. The new text follows ISO 8601 / RFC 3389.

Reviewer: But ISO 8601 / RFC 3389 don't really apply to calendars on idealised worlds, I guess, so strict adherence in the case of perpetual calendar and no calendar isn't really necessary.

Author: But then the whole year can be deleted because it has no meaning. Inversely, any year could be indicated. The ISO 8601 adherence is more about consistent formatting than anything else, seeing that most CF calendars do not agree with the standard anyway.

Reviewer: The year does have a meaning within the context of the experiment, and it might not be arbitrary because it could have been agreed with others in a comparison project. For example, if it was a very long low-resolution idealised AOGCM experiment on the hysteresis of AMOC, it might start from year 1 and run for more than 1e4 years. If it was a paleoclimate experiment for the Last Glacial Maximum, it might start from year -26e3, or -120e3 for the Last Interglacial. These are meaningful years because of orbital forcing. In both cases, the year numbers exceed four digits, and there's no logic in padding to four digits.

Reviewer 2: Just as a side note, intended as a heads up and not to interfere with moving this PR forward to close the associated issue: 

The 5th paragraph of the [CF 1.4 Overview](https://cfconventions.org/Data/cf-conventions/cf-conventions-1.13/cf-conventions.html#overview) states that 
> The values of the units attributes are character strings that are recognized by UNIDATA’s UDUNITS package

However, the current version of UDUNITS  does not allow the years part to have more than 4 digits. Work is ongoing to change this.

Author: Dear Jonathan, in the examples you give wouldn't it be more appropriate to use a regular model calendar rather than `none` which is for "climate experiments that simulate a fixed time of year"? In either case, the year has _at least_ four digits but there may be more, as per ISO 8601.

Reviewer 2: Now actually commenting on the PR, I think that it is highly relevant to require at least 4 digits for the year part. Given the general confusion of the order of year, month, and day in all the different alternatives that are in use, it is not at all clear if 1-2-3 means 0001-FEB-03, 0003-FEB-01, 0001-MAR-02, or 0003-JAN-02 (hopefully the two 0002-\*-\* are less likely). While the CF interpretation of 1-2-3 of course the first alternative, it is not unlikely that persons used to any of the alternatives may make mistakes, either when writing or when reading. Thus, requiring at least four digits for year will reduce the possibilities for mistakes. 
@pvanlaake : It is generally not possible/easy to change a model's calendar, typically require substantially model development.

Reviewer: I thought this would be a minor comment on the text, rather than a debate of principle, which I should have put in the issue instead! Thanks for your replies. I accept your arguments and withdraw my initial comment.

----
double time(time) ;
  time:long_name = "time" ;

>   time:units = "days since 0001-01-01" ;

Reviewer: I think we should leave the year as `1` because we can't know how many leading zeroes would be appropriate in this idealised calendar [same as previous comment].

Author: See reply to previous comment
